rllib/core/learner/tests/test_learner.py

```diff
@@ -225,109 +232,97 @@ def test_add_remove_module(self):
-
-        # remove a module then save/load and check states
-        with tempfile.TemporaryDirectory() as tmpdir:
-            learner1.remove_module(module_id=DEFAULT_POLICY_ID)
-            learner1.save_state(tmpdir)
-            learner2.load_state(tmpdir)
-            self._check_learner_states(learner1, learner2)
-
-    def _check_learner_states(self, learner1, learner2):
+        """Tests, whether a Learner's state is properly saved and restored."""
+        for fw in framework_iterator(frameworks=("torch", "tf2")):
+            #
+            learner1 = get_learner(framework=fw, env=self.ENV)
+            with tempfile.TemporaryDirectory() as tmpdir:
+                learner1.save_state(tmpdir)
+
+                learner2 = get_learner(framework=fw, env=self.ENV)
+                learner2.load_state(tmpdir)
+                self._check_learner_states(fw, learner1, learner2)
+
+            # add a module then save/load and check states
+            with tempfile.TemporaryDirectory() as tmpdir:
+                learner1.add_module(
+                    module_id="test",
+                    module_spec=get_module_spec(framework=fw, env=self.ENV),
+                )
+                learner1.save_state(tmpdir)
+                learner2.load_state(tmpdir)
+                self._check_learner_states(fw, learner1, learner2)
+
+            # remove a module then save/load and check states
+            with tempfile.TemporaryDirectory() as tmpdir:
+                learner1.remove_module(module_id=DEFAULT_POLICY_ID)
+                learner1.save_state(tmpdir)
+                learner2.load_state(tmpdir)
+                self._check_learner_states(fw, learner1, learner2)
+
+    def _check_learner_states(self, framework, learner1, learner2):
         check(learner1.get_weights(), learner2.get_weights())
 
+        # Method to call on the localk optimizer object to get the optimizer's
+        # state.
+        method = "get_config" if framework == "tf2" else "state_dict"
+
         # check all internal optimizer state dictionaries have been updated
         learner_1_optims_serialized = {
-            name: optim.get_config()
+            name: getattr(optim, method)()
             for name, optim in learner1._named_optimizers.items()
         }
         learner_2_optims_serialized = {
-            name: optim.get_config()
+            name: getattr(optim, method)()
             for name, optim in learner2._named_optimizers.items()
         }
         check(learner_1_optims_serialized, learner_2_optims_serialized)
 
         learner_1_optims_serialized = [
-            optim.get_config() for optim in learner1._optimizer_parameters.keys()
+            getattr(optim, method)() for optim in learner1._optimizer_parameters.keys()
         ]
         learner_2_optims_serialized = [
-            optim.get_config() for optim in learner2._optimizer_parameters.keys()
+            getattr(optim, method)() for optim in learner2._optimizer_parameters.keys()
```

Comment:
These lines here, what are they doing that the previous lines are not doing? Should we get rid of them? one is checking the values in the list while the previous one is checking them as a dict. The dict one is more comprehensive.

Reply:
Not sure I understand. They check that the updates applied to an optimizer have the same effect, after you have restored the optimizer(s) from a checkpoint. This is not a new check I'm adding here, I'm just making this test framework-independent.